docs/developers/components/README.md

```diff
@@ -1,34 +1,95 @@
 
-Server components can be easily be added and configured by simply adding an entry to your harperdb-config.yaml:
+Functionally, what makes an Extension a Component is the contents of `config.yaml`. Unlike the Application Template referenced earlier, which specified multiple Components within the _config.yaml_, an Extension will only specify `extensionModule: <pathToExtension>`. The path must resolve from the root of the Extension module directory.
 
-```yaml
-my-server-component:
-  package: 'HarperDB-Add-Ons/package-name' # this can be any valid github or npm reference
-  port: 4321
-```
+For example, the [Harper Next.js Extension]() _config.yaml_ specifies `extensionModule: ./extension.js`.
 
-## Writing Extension Components
+It is also recommended that all Extensions have a _package.json_ that specifies JavaScript package metadata such as name, version, etc. Since Extensions are just JavaScript packages, they can do anything a JavaScript package can normally do. It can be written in TypeScript, and compiled to JavaScript. It can export an executable (using the [bin]() property). It can be published to npm. The possibilities are endless!
 
-You can write your own extensions to build new functionality on HarperDB. See the [writing extension components documentation](writing-extensions.md) for more information.
+<!-- TODO: Write section here about Worker Threads -->
+
+### Resource Extension
+
+A Resource Extension is for processing a certain type of file or directory. For example, the internal [jsResource]() extension handles executing JavaScript files.
+
+These Extensions are comprised of four distinct function exports, `handleFile()`, `handleDirectory()`, `setupFile()`, and `setupDirectory()`. The `handleFile()` and `handleDirectory()` methods are executed on **all worker threads**, and are _executed again during restarts_. The `setupFile()` and `setupDirectory()` methods are only executed **once** on the **main thread** during the initial system start sequence.
+
+> Keep in mind that the CLI command `harperdb restart` or CLI argument `restart=true` only restarts the worker threads. If a Component is deployed using `harperdb deploy`, the code within the `setup{File|Directory}()` methods will not be executed until the system is completely shutdown and turned back on.
+
+Other than their execution behavior, the `{handle|setup}File()` and `{handle|setup}Directory()` methods have identical function definitions (arguments and return value behavior).
+
+#### `{handle|setup}File(contents, urlPath, path, resources): void | Promise<void>`
+
+This method is for processing individual files. It can be async.
+
+> Remember!
+> 
+> `setupFile()` is executed **once** on the **main thread** during the main start sequence.
+> 
+> `handleFile()` is executed on **worker threads** and is executed again during restarts.
+
+Parameters:
+
+- **contents** - `Buffer` - The contents of the file
+- **urlPath** - `String` - ???
```

Comment:
This the recommended URL path that the extension should use for any resources that it registers.